File: GGAndroid_2.0/GGCore/GGRxJava/src/main/java/io/ganguo/rx/RxCollections.java
package io.ganguo.rx;


import android.support.annotation.NonNull;

import java.util.Collection;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableTransformer;
import io.reactivex.functions.Function;

/**
 * 集合类常用操作符
 * Created by Roger on 4/29/16.
 */
public class RxCollections {

    /**
     * 如果集合不为空，发射原数据集合，否则发射一个ObservableEmpty
     *
     * @param <T>
     * @return
     */
    public static <T extends Collection> ObservableTransformer<T, T> filterNotEmpty() {
        return new ObservableTransformer<T, T>() {
            @Override
            public Observable<T> apply(Observable<T> upStream) {
                return upStream.compose(RxStatement.ifThen(
                        new Condition<T>() {
                            @Override
                            public boolean call(T collection) {
                                return isEmpty(collection);
                            }
                        },
                        new Mapper<T, Observable<T>>() {
                            @Override
                            public Observable<T> map(T t) {
                                return Observable.<T>empty();
                            }
                        }
                ));
            }
        };
    }

    public static <E, T extends Collection<E>> ObservableTransformer<T, E> emitItems() {
        return new ObservableTransformer<T, E>() {
            @Override
            public Observable<E> apply(Observable<T> tObservable) {
                return tObservable.compose(RxCollections.<T>filterNotEmpty())
                        .flatMap(new Function<T, Observable<E>>() {
                            @Override
                            public Observable<E> apply(T collection) {
                                return Observable.fromIterable(collection);
                            }
                        });
            }
        };
    }

    /**
     * Map data to List and emit items
     *
     * @param dataMapper 转换 upStream data 为 List 类型
     * @param <T>        upStream 数据类型
     * @param <E>        发射的 item 数据类型
     * @return
     */
    public static <T, E> ObservableTransformer<T, E> emitItems(@NonNull final Mapper<T, Collection<E>> dataMapper) {
        return new ObservableTransformer<T, E>() {
            @Override
            public Observable<E> apply(Observable<T> upStream) {
                return upStream
                        .flatMap(new Function<T, Observable<E>>() {
                            @Override
                            public Observable<E> apply(T t) {
                                Collection<E> collection = dataMapper.map(t);
                                if (!isEmpty(collection)) {
                                    return Observable.fromIterable(collection);
                                } else {
                                    return Observable.empty();
                                }
                            }
                        });
            }
        };
    }

    /**
     * 转换原 List 为另一种 List 类型
     */
    /**
     * @param convertor 转换器
     * @param <T>       原 List item 数据类型
     * @param <E>       转换后的 List item 数据类型
     * @return
     */
    public static <T, E> ObservableTransformer<Iterable<T>, List<E>> cast(@NonNull final Mapper<T, E> convertor) {
        return new ObservableTransformer<Iterable<T>, List<E>>() {
            @Override
            public Observable<List<E>> apply(Observable<Iterable<T>> upStream) {
                return upStream
                        .flatMap(new Function<Iterable<T>, Observable<List<E>>>() {
                            @Override
                            public Observable<List<E>> apply(Iterable<T> ts) {
                                return Observable.fromIterable(ts)
                                        .map(new Function<T, E>() {
                                            @Override
                                            public E apply(T t) throws Exception {
                                                return convertor.map(t);
                                            }
                                        })
                                        .toList()
                                        .toObservable();
                            }
                        });
            }
        };
    }

    public static boolean isEmpty(Collection<?> collection) {
        return collection == null || collection.isEmpty();
    }

    public static boolean isNotEmpty(Collection<?> collection) {
        return !isEmpty(collection);
    }
}

File: GGAndroid_3.0/core/support/rxjava/src/main/java/io/ganguo/rx/utils/RxCollections.kt
package io.ganguo.rx.utils

import io.ganguo.rx.transformer.RxTransformer
import io.ganguo.rx.transformer.interceptor.RxInterceptor
import io.ganguo.rx.transformer.interceptor.RxObservableInterceptor
import io.ganguo.rx.transformer.transfer.RxObservableTransfer
import io.ganguo.rx.transformer.transfer.RxTransfer
import io.reactivex.Observable
import io.reactivex.functions.Function

/**
 * <pre>
 *     author : leo
 *     time   : 2020/05/12
 *     desc   : RxJava Collections Transformer
 * </pre>
 */


/**
 * Filter for an empty collection
 * @param <T>
 * @return
 */
fun <T> filterNotEmpty(): RxInterceptor<Collection<T>> {
    return RxTransformer.warp(object : RxObservableInterceptor<Collection<T>> {
        override fun applyInterceptor(observable: Observable<Collection<T>>): Observable<Collection<T>> {
            return observable
                    .flatMap {
                        if (it.isEmpty()) {
                            Observable.empty()
                        } else {
                            Observable.just(it)
                        }
                    }
        }

    })
}


/**
 * Takes a single item from the collection
 * @return RxTransfer<T, E>
 */
fun <E> emitItems(): RxTransfer<Collection<E>, E> {
    return RxTransformer.warp(object : RxObservableTransfer<Collection<E>, E> {
        override fun applyTransfer(observable: Observable<Collection<E>>): Observable<E> {
            return observable
                    .compose(filterNotEmpty())
                    .flatMap {
                        Observable.fromIterable(it)
                    }
        }

    })
}

/**
 * Map data to List and emit items
 * @param dataMapper Convert observable data to type List
 * @return
 */
fun <T, E> emitItems(dataMapper: Function<T, Collection<E>?>): RxTransfer<T, E> {
    return RxTransformer.warp(object : RxObservableTransfer<T, E> {
        override fun applyTransfer(observable: Observable<T>): Observable<E> {
            return observable
                    .flatMap {
                        val collection = dataMapper.apply(it)
                        if (!collection.isNullOrEmpty()) {
                            Observable.fromIterable(collection)
                        } else {
                            Observable.empty()
                        }
                    }
        }
    })
}

/**
 * List<T> convert to List<E>
 * @param convert Function<T, E>
 * @return RxTransfer<Iterable<T>, List<E>>
 */
fun <T, E> cast(convert: Function<T, E>): RxTransfer<Iterable<T>, List<E>> {
    return RxTransformer.warp(object : RxObservableTransfer<Iterable<T>, List<E>> {
        override fun applyTransfer(observable: Observable<Iterable<T>>): Observable<List<E>> {
            return observable
                    .flatMap {
                        Observable.fromIterable(it)
                    }
                    .map {
                        convert.apply(it)
                    }
                    .toList()
                    .toObservable()
        }
    })
}
